### ISSUE TITLE
DWT on Linux with Wayland?

### ISSUE THREAD

**Issue by russel** [d-widget-toolkit/dwt#42]
The list of dependencies for using DWT on GTK+ appear to imply Xorg as the server. Does DWT work with Wayland?

**Comment by russel** [comment]
Debian Sid seems to have packages relating to SWT 3.8.2 and 4.7.3. The documentation implies you have to start a JVM to use SWT, which would seem to make DWT only indirectly a native system – having to start JVM would be a real pain.

**Comment by jacob-carlborg** [comment]
DWT is a **port** of SWT, not a binding. The full source code of SWT has been translated from Java to D. The thin C/JNI layer has been removed since it's not needed in D. Not other dependencies than system libraries are required.

**Comment by jacob-carlborg** [comment]
> Interesting question then as to which Debian Sid packages to depend on.

The list in the readme [1] is based on Ubuntu. I would guess the packages for Debian would be the same (or very similar).

[1] https://github.com/d-widget-toolkit/dwt#usage

**Comment by russel** [comment]
I believe these packages are just taken from Debian by Ubuntu. All with libx… and any with x11 in the name indicate a direct dependency on X11. Ubuntu has backed off going with Wayland in the last couple of releases, but Debian and Fedora have made Wayland the default. You can still login using X11 but Wayland is the default. Ubuntu will eventually have to relent and make Wayland their default, SWT people will have to think about what to do with all their X11 dependencies. 

It is true that Xwayland incorporates a lot of the X11 protocols, but it also removes a lot of the X11 graphics. Gromit-MPX for example works on Xorg but not on Xwayland. If SWT doesn't use these then things will be OK. If however it does, then SWT will need to change and quickly. I wonder if Eclipse works on Xwayland? The JetBrains IDEs do.

**Comment by russel** [comment]
I believe Ubuntu just pulls these packages from Debian unchanged. With so many libx.* and .*x11.* packages the dependency on X11 is very clear. I guess though it depends on whether they use things that are in Xorg but not in Xwayland. Xwayland supports a lot of the X11 protocol (where Wayland does not) but not all the graphics stuff. Gromit-MPX works on Xorg but fails on Xwayland for example.

I guess Eclipse has to work on Xwayland so perhaps SWT is already ready.

Currently Ubuntu runs GNOME on Xorg by default where Debian and Fedora runs GNOME on Xwayland by default, but you can use Xorg, it's just not the default. Ubuntu will have to switch its default soon I suspect. 

I am feeling that getting DWT to SWT 4.7.3 and trying it on Xwayland without all the X11 dependencies is the only way forward.

**Comment by jacob-carlborg** [comment]
Ah, ok, you're referring to using Wayland. As I mentioned above, as far as I could see SWT added support for Wayland in version 4.7. DWT is based on SWT 3.4 so it's quite far behind.

**Comment by jacob-carlborg** [comment]
> I am feeling that getting DWT to SWT 4.7.3 and trying it on Xwayland without all the X11 dependencies is the only way forward. 

Definitely. I’m working on a tool that will automatically translate the Java code to D which should help with this problem.

**Comment by jacob-carlborg** [comment]
Some work has started to update DWT to a later version of SWT.
